### DISCUSSION TITLE
How do I provide a file as a dependency to an external crate's build.rs?

### DISCUSSION THREAD

**Discussion by akesling** [bazelbuild/rules_rust#2552] [Q&A] [answered]
I'm trying to link in `pyo3` with a custom build of Python and a pyo3 build config file.  This is normally passed as an absolute path via the `PYO3_CONFIG_FILE` environment variable.  I've figured out [how to pass the appropriate environment variable](https://github.com/bazelbuild/rules_rust/discussions/2532), but I don't yet know the most ergonomic way to both A) make a file available at build script execution time to the appropriate crate (`pyo3-ffi` in this case), and B) set `PYO3_CONFIG_FILE` to an absolute path which points at said provided file.

Any ideas?

For context, the cargo command I'd use alternatively would look something like: `PYO3_CONFIG_FILE="$(pwd)/pyo3_build_config.txt" PYO3_NO_PYTHON=1 cargo build --release --bin runtime -p python-embedder`

**Comment by illicitonion** [comment]
You probably need to add the file as a label to `build_script_data` annotation attribute for the crate, and then you can use `$(execpath //that:label)` in the value of `build_script_env` on the annotation to get an absolute path to that file.

**Reply by akesling** [reply]
@illicitonion thank you for the pointers!

Here's what I'm trying now:
```starlark
...
    annotations = {
        # Set pyo3 FFI build flags when building pyo3-ffi dep
        "pyo3-ffi": [crate.annotation(
            build_script_data = ["//python_embedder:pyo3_build_config.txt"],
            build_script_env = {
                "PYO3_NO_PYTHON": "1",
                "PYO3_CONFIG_FILE": "$(execpath //python_embedder:pyo3_build_config.txt)",
            },
        )],
    },
...
```

But when I build, I get the following: `ERROR: /private/var/tmp/_bazel_akesling/72b5820b02a477ec8ed97bd441726217/external/crates__pyo3-ffi-0.19.2/BUILD.bazel:90:19: no such package '@crates__pyo3-ffi-0.19.2//python_embedder': BUILD file not found in directory 'python_embedder' of external repository @crates__pyo3-ffi-0.19.2. Add a BUILD file to a directory to mark it as a package. and referenced by '@crates__pyo3-ffi-0.19.2//:pyo3-ffi_build_script'
`.

It looks like the path is being interpreted as being relative to some `pyo3-ffi` root and not my repo.   Any advice here?

**Reply by illicitonion** [reply]
Try throwing in a `@` before the `//` for both labels - `@//` means "in the root repo":

```starlark
...
    annotations = {
        # Set pyo3 FFI build flags when building pyo3-ffi dep
        "pyo3-ffi": [crate.annotation(
            build_script_data = ["@//python_embedder:pyo3_build_config.txt"],
            build_script_env = {
                "PYO3_NO_PYTHON": "1",
                "PYO3_CONFIG_FILE": "$(execpath @//python_embedder:pyo3_build_config.txt)",
            },
        )],
    },
...
```

**Reply by akesling** [reply, marked as answer]
Thank you so much for the help!

To close the loop here (in case someone stumbles across this looking for how to actually get this working w/ PyO3):

To get the equivalent of `PYO3_CONFIG_FILE="$(pwd)/pyo3_build_config.txt" PYO3_NO_PYTHON=1 cargo build --release --bin runtime -p python-embedder`, the following annotations work....

```starlark
...
    annotations = {
        "pyo3-build-config": [crate.annotation(
            build_script_data = ["@//python_embedder:pyo3_build_config.txt"],
            build_script_env = {
                "PYO3_NO_PYTHON": "1",
                "PYO3_CONFIG_FILE": "$(execpath @//python_embedder:pyo3_build_config.txt)",
            },
        )],
    },
...
```

Note that we're directly using the `pyo3` crate, but we fundamentally depend on `pyo3-build-config` to consume the PyO3 build configuration.
